**Exercício 1: Cálculo de Distância Euclidiana**<br>
Descrição: Escreva uma função em Python que receba dois vetores (listas de números) de mesma dimensão e retorne a distância euclidiana entre eles.

In [ ]:
import math

def euclidean_distance(vector1, vector2):
   """
    Calcula a distância euclidiana entre dois vetores.

     A distância euclidiana é a "linha reta" entre dois pontos no espaço.
     Fórmula: d = √[(x₁-x₂)² + (y₁-y₂)² + ... + (n₁-n₂)²]

     Args:
         vetor1: Lista de números (primeiro vetor)
         vetor2: Lista de números (segundo vetor)

     Returns:
         float: Distância euclidiana entre os vetores

     Raises:
         ValueError: Se os vetores têm dimensões diferentes

   """
   # Verificar se os vetores têm a mesma dimensão
   if len(vector1) != len(vector2):
      raise ValueError(f"Vetors must have the same dimension"
                       f"Vector1: {len(vector1)}, Vector2: {len(vector2)}")
   
   # Calcular a soma dos quadrados das diferenças
   sum_squares = sum((a - b)**2 for a, b in zip(vector1, vector2))

   return math.sqrt(sum_squares)

# Exemplo de Uso

# Teste Básico
v1 = [0, 0]
v2 = [3, 4]
print(f"Distance between {v1} and {v2}: {euclidean_distance(v1, v2)}")

# Teste 3D
v3 = [1, 2, 3]
v4 = [4, 5, 6]
print(f"Distance between {v3} and {v4}: {euclidean_distance(v3, v4)}")

# Teste com decimais
v5 = [1.5, 2.5]
v6 = [3.0, 1.0]
print(f"Distance between {v5} and {v6}: {euclidean_distance(v5, v6)}")

# Teste interativo
print("\n TEST YOUR OWN VECTORS")
try:
   input1 = input("Enter the vector 1 (ex: 1 2 3): ").split()
   input2 = input("Enter the vector 2 (ex: 4 5 6): ").split()

   vector1 = [float(x) for x in input1]
   vector2 = [float(x) for x in input2]

   result = euclidean_distance(vector1, vector2)
   print(f"Distance: {result:.4f}")
except:
   print("Input error. Use numbers separated by spaces.")

Distance between [0, 0] and [3, 4]: 5.0
Distance between [1, 2, 3] and [4, 5, 6]: 5.196152422706632
Distance between [1.5, 2.5] and [3.0, 1.0]: 2.1213203435596424

 TEST YOUR OWN VECTORS
Distance: 5.1962


## 📐 **Análise do Código de Distância Euclidiana**

### **🌍 Contexto Geométrico**

A distância euclidiana representa a **"linha reta"** entre dois pontos no espaço, sendo a forma mais natural de medir proximidade:

- **2D (Plano)**: Como no exemplo `[0,0]` e `[3,4]`, a distância é 5.0 - exatamente a hipotenusa do triângulo retângulo 3-4-5
- **3D (Espaço)**: Vetores como `[1,2,3]` e `[4,5,6]` representam pontos no espaço tridimensional, onde a distância é a linha reta conectando-os
- **Multidimensional**: Embora não visualizável, o conceito se estende para qualquer número de dimensões

### **🔍 Similaridade**

A distância euclidiana é uma **métrica fundamental de similaridade**:

- **Menor distância = Maior similaridade**: Quanto menor o valor, mais próximos/similares são os vetores
- **Zero = Identidade**: Distância 0.0 significa vetores idênticos
- **Aplicações práticas**:
  - **Pessoas**: Comparar características físicas (altura, peso, idade)
  - **Produtos**: Avaliar similaridade entre itens
  - **Machine Learning**: Classificação por vizinhança (k-NN)
  - **Recomendações**: Encontrar usuários com gostos similares

### **📊 Interpretação Numérica**

Os resultados numéricos do código revelam padrões importantes:

1. **Exemplo 2D**: `[0,0]` → `[3,4]` = **5.0**
   - Interpretação: Distância clássica do teorema de Pitágoras
   - Significa que os pontos estão 5 unidades separados no plano

2. **Exemplo 3D**: `[1,2,3]` → `[4,5,6]` = **≈5.196**
   - Interpretação: Cada dimensão contribui igualmente (diferença de 3 em cada)
   - √(3² + 3² + 3²) = √27 ≈ 5.196

3. **Exemplo Decimal**: `[1.5,2.5]` → `[3.0,1.0]` = **≈2.121**
   - Interpretação: Demonstra que a função funciona com qualquer precisão numérica
   - √((3.0-1.5)² + (1.0-2.5)²) = √(2.25 + 2.25) = √4.5 ≈ 2.121

### **💡 Implicações Práticas**

- **Escala importa**: Características com valores maiores (ex: salário vs idade) podem dominar o cálculo
- **Normalização**: Em aplicações reais, é comum normalizar os dados primeiro
- **Interpretação contextual**: Uma distância de 5.0 pode ser pequena ou grande dependendo do contexto dos dados

O código implementa essa métrica fundamental de forma **simples e eficaz**, tornando-a aplicável desde geometria básica até análise de dados complexos.

**Exercício 2: Distância Euclidiana com Validação de Dados** <br>
Descrição: Modifique a função do Exercício 1 para validar se os dois vetores têm o mesmo tamanho. Se não tiverem, a função deve retornar um erro.

In [11]:
import math

def euclidean_distance_2(vector1, vector2):
    '''
    Calcula a distância euclidiana entre dois vetores com validação de dados.
    
    A distância euclidiana é a "linha reta" entre dois pontos no espaço.
    Fórmula: d = √[(x₁-x₂)² + (y₁-y₂)² + ... + (n₁-n₂)²]
    
    Args:
        vetor1: Lista de números (primeiro vetor)
        vetor2: Lista de números (segundo vetor)
    
    Returns:
        float: Distância euclidiana entre os vetores
    
    Raises:
        ValueError: Se os vetores têm dimensões diferentes ou estão vazios
        TypeError: Se os elementos não são números
   
   '''
    # Validação 1: Verificar se os vetores não estão vazios
    if not vector1 or not vector2:
        raise ValueError("Vectors cannot be empty!")

    # Validação 2: Verificar se os vetores têm a mesma dimensão
    if len(vector1) != len(vector2):
        raise ValueError(
            f"Vectors must have the same dimension!"
            f"Vector 1: {len(vector1)}, Vector 2: {len(vector2)}"
        )

    # Validação 3: Verificar se todos os elementos são números
    try:
        # Tentar converter para float para garantir que são números
        v1_nums = [float(x) for x in vector1]
        v2_nums = [float(x) for x in vector2]
    except (ValueError, TypeError):
        raise TypeError("All elements of the vectors must be numbers")

    sum_squares = sum((a-b)**2 for a, b in zip(v1_nums, v2_nums))

    return math.sqrt(sum_squares)

v1 = [0, 0]
v2 = [3, 4]
print(f"Distance between {v1} and {v2}: {euclidean_distance_2(v1, v2)}")

v3 = [1, 2, 3]
v4 = [4, 5, 6]
print(f"Distance between {v3} and {v4}: {euclidean_distance_2(v3, v4)}")

v5 = [1.5, 2.5]
v6 = [3.0, 1.0]
print(f"Distance between {v5} and {v6}: {euclidean_distance_2(v5, v6)}")

# Testes de validação (mostram os erros)
print("\n Testing Validations: ")

# Erro: Dimensões diferentes
try:
    result = euclidean_distance_2([1, 2], [1, 2, 3])
except ValueError as e:
    print(f"Error Caight: {e}")

# Erro: Vetor vazio
try:
    result = euclidean_distance_2([], [1, 2])
except ValueError as e:
    print(f"Error Caught: {e}")

# Erro: Elementos não numéricos
try:
    result = euclidean_distance_2([1, "a"], [2, 3])
except TypeError as e:
    print(f"Error Caught: {e}")

# Teste interativo
print("\n Testing your own arrays: ")
try:
    input1 = input("Enter the vector 1 (ex: 1 2 3): ").split()
    input2 = input("Enter the vector 2 (ex: 4 5 6): ").split()

    vector1 = [float(x) for x in input1]
    vector2 = [float(x) for x in input2]

    result = euclidean_distance_2(vector1, vector2)
    print(f"Distance: {result:.4f}")

except ValueError as e:
    print(f"Validation Erro: {e}")
except Exception as e:
    print(f"Input Error: {e}")

Distance between [0, 0] and [3, 4]: 5.0
Distance between [1, 2, 3] and [4, 5, 6]: 5.196152422706632
Distance between [1.5, 2.5] and [3.0, 1.0]: 2.1213203435596424

 Testing Validations: 
Error Caight: Vectors must have the same dimension!Vector 1: 2, Vector 2: 3
Error Caught: Vectors cannot be empty!
Error Caught: All elements of the vectors must be numbers

 Testing your own arrays: 
Distance: 5.1962


**Exercício 3: Distância Euclidiana em N-Dimensões**<br>
Descrição:  Generalize  a  função  do  Exercício  1  para  que  ela  possa  calcular  a  distância euclidiana entre dois vetores em um espaço de qualquer número de dimensões.

In [2]:
import math

def distancia_euclidiana_nd(vetor1, vetor2):
   """
   Calcula a distância euclidiana entre dois vetores em N-dimensões.
    
    A distância euclidiana generalizada para qualquer número de dimensões:
    d = √[(x₁-y₁)² + (x₂-y₂)² + ... + (xₙ-yₙ)²]
    
    Args:
        vetor1: Lista de números (primeiro vetor de N dimensões)
        vetor2: Lista de números (segundo vetor de N dimensões)
    
    Returns:
        float: Distância euclidiana entre os vetores
    
    Raises:
        ValueError: Se os vetores têm dimensões diferentes ou estão vazios
        TypeError: Se os elementos não são números
   """
   # Validação 1: Verificar se os vetores não estão vazias
   if not vetor1 or not vetor1:
      raise ValueError("Os vetores não podem estar vazios!")
   
   # Validação 2: Verificar se os vetores têm a mesma dimensão
   if len(vetor1) != len(vetor2):
      raise ValueError(f"Vetores devem ter a mesma dimensão!"
                       f"Vetor1: {len(vetor1)}, Vetor2: {len(vetor2)}")
   
   # Validação 3: Verificar se todos os elementos são números
   try:
      v1_nums = [float(x) for x in vetor1]
      v2_nums = [float(x) for x in vetor2]
   except (ValueError, TypeError):
      raise TypeError("Todos os elementos dos vetores devem ser números.")
   
   soma_quadrados = sum((a-b)**2 for a, b in zip(v1_nums, v2_nums))
   return math.sqrt(soma_quadrados)

# Exemplos de uso

# 1D - Distância na linha
print(f"1D: {distancia_euclidiana_nd([5], [12])}")

# 2D - Distância no plano
print(f"2D: {distancia_euclidiana_nd([0, 0], [3, 4])}")

# 3D - Distância no espaço
print(f"3D: {distancia_euclidiana_nd([1, 2, 3], [4, 5, 6])}")

# 5D - Exemplo prático
pessoa1 = [175, 70, 25, 50, 16] # altura, peso, idade, salário, estudos
pessoa2 = [180, 75, 30, 60, 18]
print(f"5D: {distancia_euclidiana_nd(pessoa1, pessoa2)}")

# 10D - Alta dimensionalidade
v1_10d = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
v2_10d = [2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
print(f"10D: {distancia_euclidiana_nd(v1_10d, v2_10d)}")

1D: 7.0
2D: 5.0
3D: 5.196152422706632
5D: 13.379088160259652
10D: 3.1622776601683795


**Exercício 4: Comparando Distâncias**<br>
Descrição:  Escreva  um  script  que  receba  três  vetores  de  mesma  dimensão.  Calcule  a distância  euclidiana  entre  o  primeiro  e  os  outros  dois  vetores.  Imprima  qual  dos  dois vetores está mais próximo do primeiro, baseado na distância euclidiana.

In [4]:
import math

def distancia_euclidiana(vetor1, vetor2):
   """
      Calcula a distância euclidiana entre dois vetores.
    
    Args:
        vetor1: Lista de números (primeiro vetor)
        vetor2: Lista de números (segundo vetor)
    
    Returns:
        float: Distância euclidiana entre os vetores
   """

   if len(vetor1) != len(vetor2):
      raise ValueError("Vetores devem ter a mesma dimensão.")
   
   soma_quadrados = sum((a-b)**2 for a, b in zip(vetor1, vetor2))
   return math.sqrt(soma_quadrados)

def comparar_distancias(vetor_base, vetor_a, vetor_b):
   """
      Compara as distâncias euclidianas entre um vetor base e dois outros vetores.

    Calcula qual dos dois vetores (A ou B) está mais próximo do vetor base
    usando a distância euclidiana como métrica.

    Args:
        vetor_base: Lista de números (vetor de referência)
        vetor_a: Lista de números (primeiro vetor para comparação)
        vetor_b: Lista de números (segundo vetor para comparação)

    Returns:
        tuple: (vetor_mais_proximo, distancia_menor, distancia_a, distancia_b)

    Raises:
        ValueError: Se os vetores têm dimensões diferentes ou estão vazios
        TypeError: Se os elementos não são números
   """
   if not vetor_base or not vetor_a or not vetor_b:
       raise ValueError("Os vetores não podem estar vazio.")
    
   if not (len(vetor_base) == len(vetor_a) == len(vetor_b)):
      raise ValueError(f"Todos os vetores devem ter a mesma dimensão." f"Base: {len(vetor_base)}, A: {len(vetor_a)}, B: {len(vetor_b)}")
   
   try:
      for vetor in [vetor_base, vetor_a, vetor_b]:
         [float(x) for x in vetor]
   except (ValueError, TypeError):
      raise TypeError("Todos os elementos dos vetores devem ser números.")
   
   distancia_a = distancia_euclidiana(vetor_base, vetor_a)
   distancia_b = distancia_euclidiana(vetor_base, vetor_b)

   if distancia_a < distancia_b:
      vetor_mais_proximo = "A"
      distancia_menor = distancia_a
   elif distancia_b < distancia_a:
      vetor_mais_proximo = "B"
      distancia_menor = distancia_b
   else:
      vetor_mais_proximo = "Empate"
      distancia_menor = distancia_a

   return vetor_mais_proximo, distancia_menor, distancia_a, distancia_b

def imprimir_resultado(vetor_base, vetor_a, vetor_b):
   """
    Função principal que executa a comparação e imprime os resultados.
   """
   print(f"Vetor Base: {vetor_base}")
   print(f"Vetor A:    {vetor_a}")
   print(f"Vetor B:    {vetor_b}")
   print()
    
   # Fazer a comparação
   resultado, dist_menor, dist_a, dist_b = comparar_distancias(vetor_base, vetor_a, vetor_b)
    
   # Imprimir distâncias
   print(f"Distância Base → A: {dist_a:.4f}")
   print(f"Distância Base → B: {dist_b:.4f}")
   print()
    
   # Imprimir resultado
   if resultado == "Empate":
      print("🤝 Resultado: EMPATE - Ambos os vetores estão à mesma distância!")
      print(f"   Distância igual: {dist_menor:.4f}")
   else:
      print(f"🎯 Resultado: Vetor {resultado} está mais próximo do vetor base!")
      print(f"   Distância menor: {dist_menor:.4f}")
      diferenca = abs(dist_a - dist_b)
      print(f"   Diferença: {diferenca:.4f}")

# ============================================================================
# EXEMPLOS DE TESTE
# ============================================================================

print("🧮 COMPARANDO DISTÂNCIAS EUCLIDIANAS")
print("="*45)

# Teste 1: 2D - Exemplo clássico
print("\n📐 Teste 1 (2D):")
base_2d = [0, 0]
a_2d = [3, 4]  # distância = 5
b_2d = [1, 1]  # distância = √2 ≈ 1.414
imprimir_resultado(base_2d, a_2d, b_2d)

# Teste 2: 3D - Vetores no espaço
print("\n" + "="*45)
print("📦 Teste 2 (3D):")
base_3d = [1, 1, 1]
a_3d = [2, 2, 2]  # distância = √3 ≈ 1.732
b_3d = [4, 1, 1]  # distância = 3
imprimir_resultado(base_3d, a_3d, b_3d)

# Teste 3: Empate
print("\n" + "="*45)
print("🤝 Teste 3 (Empate):")
base_empate = [0, 0]
a_empate = [1, 0]  # distância = 1
b_empate = [0, 1]  # distância = 1
imprimir_resultado(base_empate, a_empate, b_empate)

# Teste 4: 5D - Dados de pessoas
print("\n" + "="*45)
print("👥 Teste 4 (5D - Pessoas):")
# [altura, peso, idade, salário, anos_estudo]
pessoa_base = [170, 70, 25, 50, 16]
pessoa_a = [175, 75, 27, 55, 18]  # Similar
pessoa_b = [160, 60, 45, 80, 12]  # Diferente
imprimir_resultado(pessoa_base, pessoa_a, pessoa_b)

# Teste 5: Teste de validação
print("\n" + "="*45)
print("🧪 Teste 5 (Validação de Erros):")
try:
    # Dimensões diferentes
    comparar_distancias([1, 2], [1, 2, 3], [1, 2])
except ValueError as e:
    print(f"❌ Erro capturado: {e}")

try:
    # Elementos não numéricos
    comparar_distancias([1, 2], [1, "a"], [1, 2])
except TypeError as e:
    print(f"❌ Erro capturado: {e}")

# Teste 6: Alta dimensionalidade
print("\n" + "="*45)
print("🚀 Teste 6 (10D):")
base_10d = [1] * 10
a_10d = [2] * 10  # distância = √10 ≈ 3.162
b_10d = [1.5] * 10  # distância = √(10 * 0.25) ≈ 1.581
imprimir_resultado(base_10d, a_10d, b_10d)

print("\n" + "="*45)
print("✅ Todos os testes concluídos!")

🧮 COMPARANDO DISTÂNCIAS EUCLIDIANAS

📐 Teste 1 (2D):
Vetor Base: [0, 0]
Vetor A:    [3, 4]
Vetor B:    [1, 1]

Distância Base → A: 5.0000
Distância Base → B: 1.4142

🎯 Resultado: Vetor B está mais próximo do vetor base!
   Distância menor: 1.4142
   Diferença: 3.5858

📦 Teste 2 (3D):
Vetor Base: [1, 1, 1]
Vetor A:    [2, 2, 2]
Vetor B:    [4, 1, 1]

Distância Base → A: 1.7321
Distância Base → B: 3.0000

🎯 Resultado: Vetor A está mais próximo do vetor base!
   Distância menor: 1.7321
   Diferença: 1.2679

🤝 Teste 3 (Empate):
Vetor Base: [0, 0]
Vetor A:    [1, 0]
Vetor B:    [0, 1]

Distância Base → A: 1.0000
Distância Base → B: 1.0000

🤝 Resultado: EMPATE - Ambos os vetores estão à mesma distância!
   Distância igual: 1.0000

👥 Teste 4 (5D - Pessoas):
Vetor Base: [170, 70, 25, 50, 16]
Vetor A:    [175, 75, 27, 55, 18]
Vetor B:    [160, 60, 45, 80, 12]

Distância Base → A: 9.1104
Distância Base → B: 38.9358

🎯 Resultado: Vetor A está mais próximo do vetor base!
   Distância menor: 9.1104

**Exercício 5: Aplicação em Data Science**<br>
Descrição: Dada uma matriz de  dados (lista de listas, onde cada sublista é um vetor de características)  e  um  vetor  de  referência,  escreva  uma  função  que  calcule  a  distância euclidiana de cada vetor na matriz até o vetor de referência. A função deve retornar uma lista de distâncias.

In [5]:
import math

def distancia_euclidiana_nd(vetor1, vetor2):
   """
   Calcula a distância euclidiana entre dois vetores.
    
    Args:
        vetor1: Lista de números (primeiro vetor)
        vetor2: Lista de números (segundo vetor)
    
    Returns:
        float: Distância euclidiana entre os vetores
   """
   if len(vetor1) != len(vetor2):
      raise ValueError("Vetores devem ter a mesma dimensão.")
   
   soma_quadrados = sum((a-b)**2 for a, b in zip(vetor1, vetor2))
   return math.sqrt(soma_quadrados)

def calcular_distancias_matriz(matriz_dados, vetor_referencia):
    """
    Calcula a distância euclidiana de cada vetor na matriz até o vetor de referência.
    
    Esta função é útil em Data Science para:
    - Encontrar amostras similares a uma referência
    - Clustering e classificação
    - Análise de similaridade
    - Detecção de outliers
    
    Args:
        matriz_dados: Lista de listas (cada sublista é um vetor de características)
        vetor_referencia: Lista de números (vetor de referência)
    
    Returns:
        list: Lista de distâncias euclidianas (float)
    
    Raises:
        ValueError: Se a matriz está vazia ou vetores têm dimensões diferentes
        TypeError: Se os elementos não são números
    """
    if not matriz_dados:
        raise ValueError("A matriz de dados não pode estar vazia!")

    if not vetor_referencia:
        ValueError("O vetor de referência não pode estar vazia!")

    dim_referencia = len(vetor_referencia)
    for i, vetor in enumerate(matriz_dados):
        if len(vetor) != dim_referencia:
            raise ValueError(f"Vetor {i} tem dimensão {len(vetor)}," f"esperado {dim_referencia}")

    try:
        [float(x) for x in vetor_referencia]
        for vetor in matriz_dados:
            [float(x) for x in vetor]
    except (ValueError, TypeError):
        raise TypeError("Todos os elementos devem ser números")

    distancias = []
    for vetor in matriz_dados:
        distancia = distancia_euclidiana(vetor, vetor_referencia)
        distancias.append(distancia)

    return distancias

# Exemplo de teste
print("📊 DATA SCIENCE - CÁLCULO DE DISTÂNCIAS EM MATRIZ")
print("=" * 55)

# Teste 1: Dataset de pessoas (2D)
print("\n👥 Teste 1: Dataset de pessoas (altura, peso)")
pessoas = [
    [170, 70],  # Pessoa 0
    [180, 80],  # Pessoa 1
    [160, 55],  # Pessoa 2
    [175, 75],  # Pessoa 3
    [165, 60],  # Pessoa 4
]
referencia_pessoa = [170, 70]  # Pessoa de referência

distancias = calcular_distancias_matriz(pessoas, referencia_pessoa)

print(f"Referência: {referencia_pessoa}")
print("Dataset:")
for i, pessoa in enumerate(pessoas):
    print(f"  Pessoa {i}: {pessoa} → Distância: {distancias[i]:.3f}")

# Encontrar mais similar
mais_similar = distancias.index(min(distancias))
print(f"\n🎯 Pessoa mais similar: {mais_similar} (distância: {min(distancias):.3f})")

# Teste 2: Dataset 3D (altura, peso, idade)
print("\n" + "=" * 55)
print("📦 Teste 2: Dataset 3D (altura, peso, idade)")
pessoas_3d = [
    [170, 70, 25],  # Jovem
    [175, 80, 45],  # Meia-idade
    [160, 60, 65],  # Idoso
    [180, 85, 30],  # Jovem adulto
    [165, 55, 22],  # Muito jovem
]
referencia_3d = [170, 70, 25]

distancias_3d = calcular_distancias_matriz(pessoas_3d, referencia_3d)

print(f"Referência: {referencia_3d}")
print("Dataset 3D:")
for i, pessoa in enumerate(pessoas_3d):
    print(f"  ID {i}: {pessoa} → Distância: {distancias_3d[i]:.3f}")

# Teste 3: Dataset de machine learning (5D)
print("\n" + "=" * 55)
print("🤖 Teste 3: Features de ML (5 características)")
features_ml = [
    [1.2, 3.4, 2.1, 0.8, 4.5],  # Amostra 0
    [1.5, 3.2, 2.3, 0.9, 4.2],  # Amostra 1
    [2.1, 2.8, 1.9, 1.2, 3.8],  # Amostra 2
    [0.9, 3.6, 2.5, 0.7, 4.8],  # Amostra 3
    [1.8, 3.0, 2.0, 1.0, 4.0],  # Amostra 4
]
referencia_ml = [1.0, 3.5, 2.0, 0.8, 4.5]

distancias_ml = calcular_distancias_matriz(features_ml, referencia_ml)

print(f"Referência: {referencia_ml}")
print("Features ML:")
for i, features in enumerate(features_ml):
    print(f"  Amostra {i}: {features} → Distância: {distancias_ml[i]:.3f}")

# Ranking por similaridade
indices_ordenados = sorted(range(len(distancias_ml)), key=lambda i: distancias_ml[i])
print(f"\n📈 Ranking por similaridade:")
for rank, idx in enumerate(indices_ordenados, 1):
    print(f"  {rank}º lugar: Amostra {idx} (dist: {distancias_ml[idx]:.3f})")

# Teste 4: Análise estatística
print("\n" + "=" * 55)
print("📊 Teste 4: Análise estatística das distâncias")
dataset_grande = [[i, i * 2, i * 0.5] for i in range(1, 21)]  # 20 amostras 3D
ref_stat = [10, 20, 5]

distancias_stat = calcular_distancias_matriz(dataset_grande, ref_stat)

print(f"Dataset: 20 amostras")
print(f"Referência: {ref_stat}")
print(f"Distância mínima: {min(distancias_stat):.3f}")
print(f"Distância máxima: {max(distancias_stat):.3f}")
print(f"Distância média: {sum(distancias_stat)/len(distancias_stat):.3f}")

# Outliers (distância > 2 * média)
media = sum(distancias_stat) / len(distancias_stat)
outliers = [i for i, d in enumerate(distancias_stat) if d > 2 * media]
print(f"Possíveis outliers: {outliers}")

print("\n✅ Todos os testes concluídos!")

📊 DATA SCIENCE - CÁLCULO DE DISTÂNCIAS EM MATRIZ

👥 Teste 1: Dataset de pessoas (altura, peso)
Referência: [170, 70]
Dataset:
  Pessoa 0: [170, 70] → Distância: 0.000
  Pessoa 1: [180, 80] → Distância: 14.142
  Pessoa 2: [160, 55] → Distância: 18.028
  Pessoa 3: [175, 75] → Distância: 7.071
  Pessoa 4: [165, 60] → Distância: 11.180

🎯 Pessoa mais similar: 0 (distância: 0.000)

📦 Teste 2: Dataset 3D (altura, peso, idade)
Referência: [170, 70, 25]
Dataset 3D:
  ID 0: [170, 70, 25] → Distância: 0.000
  ID 1: [175, 80, 45] → Distância: 22.913
  ID 2: [160, 60, 65] → Distância: 42.426
  ID 3: [180, 85, 30] → Distância: 18.708
  ID 4: [165, 55, 22] → Distância: 16.093

🤖 Teste 3: Features de ML (5 características)
Referência: [1.0, 3.5, 2.0, 0.8, 4.5]
Features ML:
  Amostra 0: [1.2, 3.4, 2.1, 0.8, 4.5] → Distância: 0.245
  Amostra 1: [1.5, 3.2, 2.3, 0.9, 4.2] → Distância: 0.728
  Amostra 2: [2.1, 2.8, 1.9, 1.2, 3.8] → Distância: 1.536
  Amostra 3: [0.9, 3.6, 2.5, 0.7, 4.8] → Distância: 0.608